### PR TITLE
lib/errors: unify multi-errors to fix printing, update error dependencies

### PR DESCRIPTION
**Pull request by bobheadxi**
## The issue

image

> Someone should formulate a plan to deprecate use of `lib/errors.Append` and use `lib/errors.Combine` instead. The issue here is that they’re not semantically equivalent: the former uses a bag of errors and the later has a primary and secondary. I think we need to tag each of the former with a banner error and use the bag as the secondary error. This might take some amount of manual vetting, there’s a number of references that exist.

Seemed simple enough...

### Tangent

I initially [started with a spike to explore removing `go-multierror` entirely](https://github.com/sourcegraph/sourcegraph/compare/spike-no-multierror). This revealed a few things:

- If anyone expects cockroach's `CombineErrors` to allow them to introspect on the underlying error via `Is`, `Unwrap`, `As` and friends, well **you can't**! https://github.com/cockroachdb/errors/issues/62 The underlying type, [`SecondaryError`](https://sourcegraph.com/github.com/cockroachdb/errors/-/blob/secondary/with_secondary.go?L25:6), completely obscures the second error, except in _certain printing situations_. 
	- This means that my (and maybe others') assumption that `CombineErrors` is like `Append` but for 2 errors instead of vararg errors **is not true**, the two are fundamentally different
	- You can use `Wrap`, but it doesn't allow you to wrap error A within error B, so deceptively `CombineErrors` seems to be the solution when it actually probably isn't
	- **A test reveals that `CombineErrors` doesn't show the second error in `.Error()`, `%s`, and possibly more** so the issue raised about `go-multierror` actually applies to all our methods of error combination
- I attempted to rewrite our own multi error type, but it turns out `Unwrap` doesn't really work the way I think it does with Cockroach errors. For example:

```go
reference := New("foo")
unwrapped := Unwrap(err)
assert.NotNil(t, err) // okay, so there's an underlying error to a raw "New" error...?
assert.True(t, Is(unwrapped, reference)) // fails! unwrapped error no longer asserts against a sentinel reference error
assert.True(t, Is(reference, unwrapped)) // well this works but semantically makes no sense because the second argument is supposed to be the reference
```

`New` comes with a stack layer that gets unwrapped on `Unwrap`, which makes using `New` errors as sentinel errors dangerous - if you unwrap an error too far the sentinel error just doesn't work anymore! So we can't depend on the behaviour of `Unwrap` to build a lightweight multierror layer on top of cockroach errors, and now I'm also unsure if I can depend on `Unwrap` to do anything 

**edit** so it turns out I was doing it wrong, see https://github.com/sourcegraph/sourcegraph/pull/31466

## Attempted solution

(this PR)

Attempts to bandaid-resolve the issue raised today at backend-crew by patching `go-multierror` to fix the main issue of "printing doesn't show all errors in multierrors". See: https://github.com/hashicorp/go-multierror/compare/master...sourcegraph:master

Migrates `CombineErrors` to proxy to `go-multierror` instead (see below)

Removes unnecessary `ErrorListFunc` re-export

## The better solution?

https://github.com/sourcegraph/sourcegraph/pull/31466

## Test plan

Unit tests

### CODE REVIEW

File: lib/errors/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// Enforce some invariants with our error libraries.
+
+func TestMultipleErrorPrinting(t *testing.T) {
```
Comment (efritz) [inline lib/errors/errors_test.go:12]: I was drafting a "how to do this" in my head recently and this was exact the best first move I could think of.

File: lib/errors/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	// Make sure all our ways of combining errors actually print them.
+
+	errFoo := New("foo")
+	errBar := New("bar")
+
+	for fn, str := range map[string]string{
+		"Append.Error":          Append(errFoo, errBar).Error(),
+		"Append.Sprintf":        fmt.Sprintf("%s", Append(errFoo, errBar)),
+		"CombineErrors.Error":   CombineErrors(errFoo, errBar).Error(),
+		"CombineErrors.Sprintf": fmt.Sprintf("%s", CombineErrors(errFoo, errBar)),
```
Comment (bobheadxi) [inline lib/errors/errors_test.go:22]: These both fail

**Comment by bobheadxi** [comment]
Another band-aid: replace `CombineErrors` with a call to `Append`

**Comment by camdencheek** [comment]
What exactly are we getting out of using `cockroachdb/errors`? Sentry interop?

**Comment by bobheadxi** [comment]
> What exactly are we getting out of using cockroachdb/errors? Sentry interop?

I'm not sure I understand the history behind the usage of `cockroachdb/errors`, but Sentry appears to be a big one. The library also automatically collects stack traces which seems to be handy, though I've yet to see it used in practice outside of Sentry. I think @jhchabran was about to enlighten be during our brief discussion but I had to hop off, maybe he can comment on this as well

**Comment by bobheadxi** [comment]
I've decided to patch `CombineErrors` to redirect to `Append` instead (https://github.com/sourcegraph/sourcegraph/pull/31431/commits/cea74a0668c2a8182b8d7f07e947b5be66cac73e)

**Comment by bobheadxi** [comment]
More bad news - upon adding the following test cases:

```go
		"Wrap Append Sprintf %s": fmt.Sprintf("%s", Wrap(Append(errFoo, errBar), "hello world")),
		"Wrap Append Sprintf %v": fmt.Sprintf("%v", Wrap(Append(errFoo, errBar), "hello world")),
```

Thing still fail, so @efritz 's example is not fixed by these changes. I'm investigating why, but it looks like at the moment so long as you implement `Unwrap` it will go all the way down and discard everything in between. I've tried implementing all the various interfaces it asks for and seemingly nothing can stop this unending madness

**Comment by efritz** [comment]
Feels like we're inventing new math.

**Comment by bobheadxi** [comment]
the better PR: https://github.com/sourcegraph/sourcegraph/pull/31466

**Pull request closed without merging**